### PR TITLE
Actor construction extension

### PR DESCRIPTION
**Pull request by JoeHegarty**
**Overview**
This splits out the ability to override actor construction into it's own extension.

**Rationale**
It's reasonable to want more than one lifetime extension but that then means the actor construction that runs is effectively random depending on the ordering of those extensions.
With the new ActorConstructionExtension the expectation is that you only have one.

### CODE REVIEW

File: actors/runtime/src/main/java/cloud/orbit/actors/Stage.java [hunk 1]
```diff
@@ -749,6 +751,17 @@ else if (extension.getAfterHandlerName() != null)
             extensions.add(lifetimeExtension);
         }
 
+        ActorConstructionExtension actorConstructionExtension = extensions.stream()
+                .filter(p -> p instanceof ActorConstructionExtension)
+                .map(p -> (ActorConstructionExtension) p)
+                .findFirst().orElse(null);
```
Comment (johnou) [inline actors/runtime/src/main/java/cloud/orbit/actors/Stage.java]: Or else new DefaultActorConstructionExtension()

File: actors/runtime/src/main/java/cloud/orbit/actors/runtime/ActorEntry.java [hunk 2]
```diff
@@ -107,7 +108,7 @@ public T getObject()
                 return Task.fromValue(null);
             }
         }
-        final Object newInstance = runtime.getFirstExtension(LifetimeExtension.class).newInstance(concreteClass);
+        final Object newInstance = runtime.getFirstExtension(ActorConstructionExtension.class).newInstance(concreteClass);
```
Comment (johnou) [inline actors/runtime/src/main/java/cloud/orbit/actors/runtime/ActorEntry.java]: Is this cached?

Comment (JoeHegarty) [inline actors/runtime/src/main/java/cloud/orbit/actors/runtime/ActorEntry.java]: No, I was actually thinking about that. But it doesn't change the existing behavior so it should be a separate PR.

File: actors/runtime/src/main/java/cloud/orbit/actors/Stage.java (hunk 1 above)
Comment (JoeHegarty) [inline actors/runtime/src/main/java/cloud/orbit/actors/Stage.java]: Fair enough, that makes sense.

File: actors/runtime/src/main/java/cloud/orbit/actors/runtime/DefaultActorConstructionExtension.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * Created by joeh on 2017-05-29.
+ */
+public class DefaultActorConstructionExtension implements ActorConstructionExtension
+{
+    @Override
+    public <T> T newInstance(final Class<T> concreteClass)
+    {
+        try
+        {
+            return concreteClass.newInstance();
```
Comment (johnou) [inline actors/runtime/src/main/java/cloud/orbit/actors/runtime/DefaultActorConstructionExtension.java]: Let's consider objensis

Comment (johnou) [inline actors/runtime/src/main/java/cloud/orbit/actors/runtime/DefaultActorConstructionExtension.java]: http://objenesis.org

Comment (ShawnTuatara) [inline actors/runtime/src/main/java/cloud/orbit/actors/runtime/DefaultActorConstructionExtension.java]: I would expect that to be a choice of the developer working with Orbit. They could explicitly define an ObjenesisActorConstructionExtension and have a dependency on it. I think it is fair enough for Orbit to have a requirement that a default no-arg constructor is available for the default.

Comment (JoeHegarty) [inline actors/runtime/src/main/java/cloud/orbit/actors/runtime/DefaultActorConstructionExtension.java]: I agree with @ShawnTuatara. As a general rule it's a requirement of Orbit that actors have a no-arg constructor. An application which wants to do something different is welcome to, but Orbit should not support that by default.
We've also made significant effort within Orbit to reduce our use of third-party dependencies so adding this without a clear benefit would go against that philosophy.
